pkg/volume/csi/csi_block.go

```diff
@@ -279,46 +200,121 @@ func (m *csiBlockMapper) MapDevice(devicePath, globalMapPath, volumeMapPath, vol
 	)
 
 	if err != nil {
-		glog.Errorf(log("blockMapper.MapDevice failed: %v", err))
-		if err := os.RemoveAll(dir); err != nil {
-			glog.Error(log("blockMapper.MapDevice failed to remove mapped dir after a NodePublish() error [%s]: %v", dir, err))
-		}
-		return err
+		glog.Errorf(log("blockMapper.publishVolumeForBlock failed: %v", err))
+		return "", err
 	}
 
-	return nil
+	return publishPath, nil
 }
 
-var _ volume.BlockVolumeUnmapper = &csiBlockMapper{}
-
-// TearDownDevice removes traces of the SetUpDevice.
-func (m *csiBlockMapper) TearDownDevice(globalMapPath, devicePath string) error {
+// SetUpDevice ensures the device is attached returns path where the device is located.
+func (m *csiBlockMapper) SetUpDevice() (string, error) {
 	if !m.plugin.blockEnabled {
-		return errors.New("CSIBlockVolume feature not enabled")
+		return "", errors.New("CSIBlockVolume feature not enabled")
 	}
+	glog.V(4).Infof(log("blockMapper.SetUpDevice called"))
 
-	glog.V(4).Infof(log("unmapper.TearDownDevice(globalMapPath=%s; devicePath=%s)", globalMapPath, devicePath))
+	// Get csiSource from spec
+	if m.spec == nil {
+		glog.Error(log("blockMapper.SetUpDevice spec is nil"))
+		return "", fmt.Errorf("spec is nil")
+	}
 
+	csiSource, err := getCSISourceFromSpec(m.spec)
+	if err != nil {
+		glog.Error(log("blockMapper.SetUpDevice failed to get CSI persistent source: %v", err))
+		return "", err
+	}
+
+	// Search for attachment by VolumeAttachment.Spec.Source.PersistentVolumeName
+	nodeName := string(m.plugin.host.GetNodeName())
+	attachID := getAttachmentName(csiSource.VolumeHandle, csiSource.Driver, nodeName)
+	attachment, err := m.k8s.StorageV1beta1().VolumeAttachments().Get(attachID, meta.GetOptions{})
+	if err != nil {
+		glog.Error(log("blockMapper.SetupDevice failed to get volume attachment [id=%v]: %v", attachID, err))
+		return "", err
+	}
+
+	if attachment == nil {
+		glog.Error(log("blockMapper.SetupDevice unable to find VolumeAttachment [id=%s]", attachID))
+		return "", errors.New("no existing VolumeAttachment found")
+	}
+
```

Comment:
Since calls to `m.stageVolumeForBlock` and `m.publishVolumeForBlock` happen successively at close proximity, I would recommend to create `csiClient` here (instead of creating an instance in each method).  Then pass the client as a parameter to these methods respectively.